File: src/pipeline-validator/src/pipeline_validator.go
package main

import (
	"encoding/json"
	"os"
)

type Job = struct {
	Name      string
	Commands  []string
	DependsOn []string `json:"depends_on"`
}

type PipelineSpecification = struct {
	Name string
	Jobs []Job
}

func findJobByName(jobs []Job, name string) Job {
	var foundJob Job

	for _, job := range jobs {
		if job.Name == name {
			foundJob = job
		}
	}

	return foundJob
}

func followDependencyCycle(
	currentJob Job,
	pipelineSpecification PipelineSpecification,
	trackedJobs []string,
) bool {

	if currentJob.DependsOn == nil || len(currentJob.DependsOn) == 0 {
		return true
	}

	for _, trackedJob := range trackedJobs {
		if trackedJob == currentJob.Name {
			return false
		}
	}

	for _, jobName := range currentJob.DependsOn {
		if jobName == currentJob.Name {
			return false
		}

		return followDependencyCycle(
			findJobByName(pipelineSpecification.Jobs, jobName),
			pipelineSpecification,
			append(trackedJobs, currentJob.Name),
		)
	}

	return true
}

func PipelineValidator(pipelineFileLocation string) string {
	var err error

	fileContent, err := os.ReadFile(pipelineFileLocation)
	if err != nil {
		panic(err)
	}

	var pipelineSpecification PipelineSpecification

	err = json.Unmarshal(fileContent, &pipelineSpecification)
	if err != nil {
		panic(err)
	}

	if pipelineSpecification.Name == "" {
		return "INVALID"
	}

	if pipelineSpecification.Jobs == nil || len(pipelineSpecification.Jobs) == 0 {
		return "INVALID"
	}

	for _, job := range pipelineSpecification.Jobs {
		if job.Name == "" {
			return "INVALID"
		}

		if len(job.Commands) == 0 {
			return "INVALID"
		}

		if !followDependencyCycle(job, pipelineSpecification, []string{}) {
			return "INVALID"
		}
	}

	return "VALID"
}

File: src/pipeline-validator/src/pipeline-validator.ts
export type PipelineValidatorParams = {
  filename: string;
};

export type PipelineValidatorReturn = "VALID" | "INVALID";

export type PipelineJob = {
  name: string;
  commands: string[];
  depends_on: string[];
};

export type PipelineContent = {
  name: string;
  jobs: PipelineJob[];
};

const validateName = (pipelineContent: PipelineContent): void => {
  if (!pipelineContent.name || pipelineContent.name === "") {
    throw new Error("Job name is required");
  }
};

const validateJobs = (pipelineContent: PipelineContent): void => {
  if (!pipelineContent.jobs || pipelineContent.jobs.length === 0) {
    throw new Error("Jobs are required");
  }

  if (pipelineContent.jobs.some((job) => !job.name)) {
    throw new Error("Jobs must have a name");
  }
};

const followDependencyCycle = (
  currentJob: PipelineJob,
  jobs: PipelineContent["jobs"],
  trackedJobs: PipelineJob["name"][],
) => {
  if (trackedJobs.includes(currentJob.name)) {
    throw new Error("Circular dependency detected");
  }

  (currentJob.depends_on ?? []).forEach((job) => {
    const followingJob = jobs.find((isJob) => isJob.name === job);

    if (!followingJob) {
      throw new Error("Dependency does not exist in jobs list");
    }

    followDependencyCycle(followingJob, jobs, [
      currentJob.name,
      ...trackedJobs,
    ]);
  });
};

const validateCircularDependency = (pipelineContent: PipelineContent): void => {
  pipelineContent.jobs.forEach((job) => {
    followDependencyCycle(job, pipelineContent.jobs, []);
  });
};

export const PipelineValidator = ({
  filename,
}: PipelineValidatorParams): PipelineValidatorReturn => {
  const fileContent: PipelineContent = require(`./${filename}`);

  try {
    validateName(fileContent);
    validateJobs(fileContent);
    validateCircularDependency(fileContent);
  } catch (error) {
    // console.error(error);

    return "INVALID";
  }

  return "VALID";
};
